modules/javafx.graphics/src/main/java/com/sun/javafx/css/BitSet.java

```diff
@@ -493,58 +534,55 @@ public void clear() {
         bits = EMPTY_SET;
     }
 
-    @Override
-    public int hashCode() {
-        int hash = 7;
-        if (bits.length > 0) {
-            for (int n = 0; n < bits.length; n++) {
-                final long mask = bits[n];
-                hash = 71 * hash + (int)(mask ^ (mask >>> 32));
-            }
-        }
-        return hash;
-    }
-
     @Override
     public boolean equals(Object obj) {
-
-        if (this == obj) {
+        // Note: overridden to provide a fast path; must still respect Set contract or it
+        // will not interact correctly with other sets; same goes for hashCode, do not
+        // override arbitrarily!
+        if (obj == this) {
             return true;
         }
-
-        if (obj == null || this.getClass() != obj.getClass()) {
-            return false;
+        if (obj instanceof BitSet<?> bitSet) {  // fast path if other is a BitSet
+            return equalsBitSet(bitSet);
```

Comment:
I suspect this will not work
BitSet<Integer> will never be equal to BitSet<String>, does it mean we have to do a per-element comparison?

Reply:
> I suspect this will not work BitSet will never be equal to BitSet, does it mean we have to do a per-element comparison?

If I understand correctly, you're pointing out that if you have two bit sets, of different element types (currently in use we have `BitSet<PseudoClassState>` and `BitSet<StyleClass>`) but they have the same bit pattern and size, that they could be considered equals.

This is a good point, and a problem that has been part of `BitSet` since its inception.

Reply:
I've fixed this problem, and while fixing it discovered another (existing) problem, which I also fixed:

It turns out that two `BitSet`s that have different internal array sizes are always considered not equal; but if I add elements to one `BitSet`, forcing its internal array to grow, but then remove them again, then the array isn't shrunk...

Added tests as well to cover these cases.